content/blog/2019-12-20-Internal-Function-Memoization/index.md

```diff
@@ -0,0 +1,175 @@
+We keep a running list of all recursive call instructions. For all instructions in the recursive function that use the result of a recursive call as an operand, we clone them and add the clone to a list of _dependents_ as they depend on the recursive calls. We clone instructions here so we can delete all instructions in the recursive function and insert our own. 
+
+#### Finding Recursive Call Offsets
+For each recursive call that we find, we look at the call argument and assert that it is a constant offset from the original function argument---i.e., of the form $(n-c)$ where $n$ is the original function argument and $c$ is a constant. We remember that $c$. 
+
+### Creating a Model for Iterative Programs to use Collected Information
+Once we have the information from the recursive program, we need to find a model such that given this information, can generate a iterative program. To determine what output we wanted, we wrote a iterative fibonacci program. Instead of writing a single transformation function $f$(Input Recursive Program) = Output Iterative Program, we break down the programs according to a layout. 
+
+Consider the layout of recursive and iterative fibonacci programs below.
+
+<img src="fiblay.png" style="max-width: 100%" >
+
+Here, we separate the base cases from the recursive calls in order to know which set of instructions will need to be in a loop, and which set will only need to be called once. It is important to note that in the black box, you can have any set of instructions to combine the recursive calls.
+
+Consider the layout of the iterative version of the same program below.
+
+<img src="memo.png" style="max-width: 100%" >
+
+Note: We keep the function header the same in order to more easily swap out invocations to fibonacci outside the function with our generated function.
+
+### New Code Generation
+To minimize changes to the overall code structure, we delete all blocks and instructions inside the original recursive function and insert our new instructions.
+
+First we add all the base case conditionals. Next, we declare and initialize all the values we need for iteration and the incrementing iterator. After that, we create the while loop and clone in the instructions dependent on the original recursive calls (e.g., the add instruction in fib(n-1) + fib(n-2)). Lastly, we add the return statementment to which the while loop exits. 
+
+Our code for the LLVM pass is available [here](https://github.com/liuhenry4428/llvm-pass-skeleton/tree/noauto).
+
+<!-- We start of by 
+#### Fibonacci
+Consider the fibonacci algorithm. To compute number $n$, you only need the numbers (n-1) and (n-2). Any additional previous calculations can be thrown away. Is it possible to infer this information when given a program, such that we can alter a program to decrease its space consumption? If so, what are the cases to make this more generalizable, beyond optimizing a fiboncci program?
```

Comment:
It seems like maybe this section was meant for the intro, not the end of the post?